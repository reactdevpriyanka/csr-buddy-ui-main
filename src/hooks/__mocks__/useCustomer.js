import customer from '__mock__/customers/nessa';

const mutate = jest.fn();

const mockUseCustomer = jest.fn(() => {
  return {
    mutate,
    error: null,
    data: customer,
    isValidating: false,
    updateCustomer: (id, data) =>
      new Promise((resolve) => {
        return resolve({ data: {} });
      }),
  };
});

mockUseCustomer.mutate = mutate;

export default mockUseCustomer;
