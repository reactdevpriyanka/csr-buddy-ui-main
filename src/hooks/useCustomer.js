import axios from 'axios';
import useSWR from 'swr';
import { useRouter } from 'next/router';
import { SNACKVARIANTS } from '@components/SnackMessage/SnackMessage';
import { useSnackbar } from 'notistack';

export default function useCustomer(customerId) {
  const router = useRouter();

  const { id } = router.query;

  const queryId = customerId ? customerId : id;

  const { data, error, mutate } = useSWR(
    queryId ? `/api/v1/customer/${queryId}` : null,
    async (url) => axios.get(url).then(({ data }) => data),
    {
      revalidateIfStale: false,
      revalidateOnFocus: false,
      revalidateOnReconnect: false,
      shouldRetryOnError: true,
    },
  );

  const { enqueueSnackbar } = useSnackbar();

  const updateCustomer = async (customerId, customerData, onSuccess) =>
    axios
      .patch(`/api/v1/customer/${customerId}`, customerData)
      .then(() => onSuccess())
      .catch(() => {
        enqueueSnackbar({
          messageHeader: 'Error',
          variant: SNACKVARIANTS.ERROR,
          messageSubheader: 'Failed to Update Customer Details',
        });
      });

  return {
    data,
    error,
    updateCustomer,
    mutate,
  };
}
