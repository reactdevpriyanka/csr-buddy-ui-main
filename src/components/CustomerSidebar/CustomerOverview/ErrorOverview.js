import { useCallback } from 'react';
import { Button } from '@material-ui/core';
import WarningIcon from '@mui/icons-material/Warning';
import useCustomer from '@/hooks/useCustomer';

const ErrorOverview = () => {
  const { mutate } = useCustomer();

  const onClick = useCallback(() => mutate(), [mutate]);

  return (
    <div>
      <WarningIcon />
      <p>{'An error has occurred while loading data for this customer'}</p>
      <Button onClick={onClick}>{'Try Again'}</Button>
    </div>
  );
};

export default ErrorOverview;
