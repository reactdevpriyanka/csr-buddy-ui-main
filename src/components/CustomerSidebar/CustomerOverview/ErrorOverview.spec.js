import { renderWrap } from '@utils';
import useCustomer from '@/hooks/useCustomer';
import ErrorOverview from './ErrorOverview';

jest.mock('@/hooks/useCustomer');

describe('<ErrorOverview />', () => {
  const render = renderWrap(ErrorOverview);

  const mutate = useCustomer.mutate;

  test('it should render an alert', () => {
    expect(render()).toMatchSnapshot();
  });

  describe('when user clicks on try again button', () => {
    test('it should refetch customer', () => {
      const { getByText } = render();
      render.trigger.click(getByText('Try Again'));
      expect(mutate).toHaveBeenCalled();
    });
  });
});
