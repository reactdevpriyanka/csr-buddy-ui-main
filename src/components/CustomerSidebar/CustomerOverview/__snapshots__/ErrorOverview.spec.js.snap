// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorOverview /> it should render an alert 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="WarningIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
          />
        </svg>
        <p>
          An error has occurred while loading data for this customer
        </p>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Try Again
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="WarningIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
        />
      </svg>
      <p>
        An error has occurred while loading data for this customer
      </p>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Try Again
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
